src/cache/advective_edmf_precomputed_quantities.jl

```diff
@@ -0,0 +1,268 @@
+#####
+##### Precomputed quantities
+#####
+import Thermodynamics as TD
+import ClimaCore: Spaces, Fields
+
+"""
+    set_advective_edmf_precomputed_quantities!(Y, p, t)
+
+Updates the precomputed quantities stored in `p` for edmfx.
+"""
+function set_advective_edmf_precomputed_quantities!(Y, p, ᶠuₕ³, t)
+    (; energy_form, moisture_model, turbconv_model) = p.atmos
+    #EDMFX BCs only support total energy as state variable
+    @assert energy_form isa TotalEnergy
+    @assert !(moisture_model isa DryModel)
+
+    FT = Spaces.undertype(axes(Y.c))
+    (; params) = p
+    (; dt) = p.simulation
+    thermo_params = CAP.thermodynamics_params(params)
+    n = n_mass_flux_subdomains(turbconv_model)
+    thermo_args = (thermo_params, energy_form, moisture_model)
+
+    (; ᶜspecific, ᶜp, ᶜΦ, ᶜh_tot, ᶜρ_ref) = p
+    (; ᶜtke⁰, ᶜρa⁰, ᶠu₃⁰, ᶜu⁰, ᶠu³⁰, ᶜK⁰, ᶜts⁰, ᶜρ⁰, ᶜh_tot⁰, ᶜq_tot⁰) = p
+    (; ᶜmixing_length, ᶜlinear_buoygrad, ᶜstrain_rate_norm, ᶜK_u, ᶜK_h) = p
+    (; ᶜuʲs, ᶠu³ʲs, ᶜKʲs, ᶜtsʲs, ᶜρʲs, ᶜentrʲs, ᶜdetrʲs) = p
+    (; ustar, obukhov_length, buoyancy_flux) = p.sfc_conditions
+
+    @. ᶜρa⁰ = ρa⁰(Y.c)
+    @. ᶜtke⁰ = divide_by_ρa(Y.c.sgs⁰.ρatke, ᶜρa⁰, 0, Y.c.ρ, turbconv_model)
+    # TODO: use all updrafts
+    @. ᶜh_tot⁰ = divide_by_ρa(
+        Y.c.ρ * ᶜh_tot - (Y.c.sgsʲs.:1).ρa * (Y.c.sgsʲs.:1).h_tot,
+        ᶜρa⁰,
+        Y.c.ρ * ᶜh_tot,
+        Y.c.ρ,
+        turbconv_model,
+    )
+    @. ᶜq_tot⁰ = divide_by_ρa(
+        Y.c.ρq_tot - (Y.c.sgsʲs.:1).ρa * (Y.c.sgsʲs.:1).q_tot,
+        ᶜρa⁰,
+        Y.c.ρq_tot,
+        Y.c.ρ,
+        turbconv_model,
+    )
+    set_sgs_ᶠu₃!(u₃⁰, ᶠu₃⁰, Y, turbconv_model)
+    set_velocity_quantities!(ᶜu⁰, ᶠu³⁰, ᶜK⁰, ᶠu₃⁰, Y.c.uₕ, ᶠuₕ³)
+    @. ᶜK⁰ += ᶜtke⁰
+    @. ᶜts⁰ = TD.PhaseEquil_phq(thermo_params, ᶜp, ᶜh_tot⁰ - ᶜK⁰ - ᶜΦ, ᶜq_tot⁰)
+    @. ᶜρ⁰ = TD.air_density(thermo_params, ᶜts⁰)
+
+    for j in 1:n
+        ᶜuʲ = ᶜuʲs.:($j)
+        ᶠu³ʲ = ᶠu³ʲs.:($j)
+        ᶜKʲ = ᶜKʲs.:($j)
+        ᶠu₃ʲ = Y.f.sgsʲs.:($j).u₃
+        ᶜtsʲ = ᶜtsʲs.:($j)
+        ᶜρʲ = ᶜρʲs.:($j)
+        ᶜh_totʲ = Y.c.sgsʲs.:($j).h_tot
+        ᶜq_totʲ = Y.c.sgsʲs.:($j).q_tot
+
+        set_velocity_quantities!(ᶜuʲ, ᶠu³ʲ, ᶜKʲ, ᶠu₃ʲ, Y.c.uₕ, ᶠuₕ³)
+        @. ᶜtsʲ =
+            TD.PhaseEquil_phq(thermo_params, ᶜp, ᶜh_totʲ - ᶜKʲ - ᶜΦ, ᶜq_totʲ)
+        @. ᶜρʲ = TD.air_density(thermo_params, ᶜtsʲ)
+
+        # When ᶜe_intʲ = ᶜe_int and ᶜq_totʲ = ᶜq_tot, we still observe that
+        # ᶜρʲ != ᶜρ. This is because the conversion from ᶜρ to ᶜp to ᶜρʲ
+        # introduces a tiny round-off error of order epsilon to ᶜρʲ. If left
+        # unchecked, this round-off error then changes the tendency of ᶠu₃ʲ,
+        # which in turn introduces an error to ᶠu³ʲ, which then increases
+        # the error in ᶜρʲ. For now, we will filter ᶜρʲ to fix this. Note
+        # that this will no longer be necessary after we add diffusion.
+        @. ᶜρʲ = ifelse(abs(ᶜρʲ - Y.c.ρ) <= 2 * eps(Y.c.ρ), Y.c.ρ, ᶜρʲ)
```

Comment:
Maybe we can remove this now as grid-mean and updraft won't be exactly the same anway